### PR TITLE
Implementation of the user-level fault handler

### PR DESCRIPTION
**Pull request by WenyuanShao**
### Summary of this Pull Request (PR)

This PR is the implementation of a user-level fault handler. The system will make a sinv call to the user level fault handler to get the fault handled when a fault or exception happens. This implementation includes the modification below.

1. Modified the comp_info and set a bit into the captbl of the comp_info on the top of the invstk to indicate whether this is a sinv call due to a fault/exception or not . Add some additional comments to make this design easier to understand.
2. Created a new type of sinv cap for the fault handlers to get rid of an extra branch in the fast path.
3. Expanded the thd_introspect to get the register information in user_level and print these register information when fault happens.
4. Create an interface for fault handlers so different components can implement fault handlers differently.

### Intent for your PR

Choose one (Mandatory):

- [ ] This PR is for a code-review and is intended to get feedback, but not to be pulled yet.
- [x] This PR is mature, and ready to be integrated into the repo.

### Reviewers (Mandatory):
(Specify @<github.com username(s)> of the reviewers. Ex: @user1, @user2)
@gparmer , @hungry-foolish , @phanikishoreg 

### Code Quality

As part of this pull request, I've considered the following:

[Style](https://github.com/gparmer/composite/raw/ppos/doc/style_guide/composite_coding_style.pdf):

- [x] Comments adhere to the Style Guide (SG)
- [x] Spacing adhere's to the SG
- [x] Naming adhere's to the SG
- [x] All other aspects of the SG are adhered to, or exceptions are justified in this pull request
- [ ] I have run the auto formatter on my code before submitting this PR (see doc/auto_formatter.md for instructions)

[Code Craftsmanship](http://www2.seas.gwu.edu/~gparmer/posts/2016-03-07-code-craftsmanship.html):

- [x] I've made an attempt to remove all redundant code
- [x] I've considered ways in which my changes might impact existing code, and cleaned it up
- [x] I've formatted the code in an effort to make it easier to read (proper error handling, function use, etc...)
- [x] I've commented appropriately where code is tricky
- [x] I agree that there is no "throw-away" code, and that code in this PR is of high quality

### Testing

I've tested the code using the following test programs (provide list here):

- [x] micro_booter
- [x] unit_pingpong
- [ ] unit_schedtests
- [x] llbooter_test

### CODE REVIEW

File: src/components/implementation/no_interface/llbooter/boot_deps.h [hunk 1]
```diff
@@ -9,10 +9,20 @@
 #include <res_spec.h>
 #include <ps.h>
 #include <bitmap.h>
+#include <fault_handler.h>
 
 /* Assembly function for sinv from new component */
 extern word_t hypercall_entry_rets_inv(spdid_t cur, int op, word_t arg1, word_t arg2, word_t arg3, word_t *ret2, word_t *ret3);
 
+/* Assembly function for sinv for the fault handlers */
```
Comment (hungry-foolish) [inline src/components/implementation/no_interface/llbooter/boot_deps.h]: consider defining unsigned longs as something else in the future, perhaps ptrint_t. This is for the long term.

File: src/components/implementation/no_interface/llbooter/Makefile [hunk 2]
```diff
@@ -1,7 +1,7 @@
 COMPONENT=llbooter.o
 ASM_OBJS=s_stub.o
 COMPONENT=llboot_comp.o
-INTERFACES=
+INTERFACES=fault_handler
```
Comment (phanikishoreg) [inline src/components/implementation/no_interface/llbooter/Makefile]: Name choices: `faults`, `faultmgr` and of course `fault_handler` here. I actually like having the API name contain interface name as a prefix but again that's a personal preference, just wanted to mention that.

File: src/components/implementation/no_interface/llbooter/boot_deps.h (hunk 1 above)
Comment (phanikishoreg) [reply src/components/implementation/no_interface/llbooter/boot_deps.h]: Why aren't they just `word_t`. That is pretty architecture specific, isn't? You're also using that for registers among others so!

File: src/components/implementation/no_interface/llbooter/boot_deps.h [hunk 3]
```diff
@@ -114,6 +124,91 @@ boot_spd_initaep_get(spdid_t spdid)
...
 
+void
+fault_reg_print(spdid_t spdid)
+{
+	struct pt_regs fault_regs;
+
+	struct cos_aep_info *child_aep = boot_spd_initaep_get(spdid);
+	struct cos_compinfo *boot_info = boot_spd_compinfo_curr_get();
+	
+	fault_regs.ax = cos_introspect(boot_info, child_aep->thd, THD_GET_FAULT_REG0);
```
[the remark this answers is not in the document]
Comment (phanikishoreg) [reply src/components/implementation/no_interface/llbooter/boot_deps.h]: What are `FAULT_REGXX`s? Are they different from the regular `REGXX`s?

File: src/components/implementation/no_interface/llbooter/boot_deps.h [hunk 4]
```diff
@@ -114,6 +124,96 @@ boot_spd_initaep_get(spdid_t spdid)
 	return cos_sched_aep_get(boot_spd_defcompinfo_get(spdid));
 }
 
+void
+fault_reg_print(spdid_t spdid)
+{
+	struct pt_regs fault_regs;
+
+	struct cos_aep_info *child_aep = boot_spd_initaep_get(spdid);
```
Comment (phanikishoreg) [inline src/components/implementation/no_interface/llbooter/boot_deps.h]: Very interesting! It looks like the handler assumes the fault is always in the `INITTHD` of that component?

File: src/components/implementation/no_interface/llbooter/boot_deps.h [hunk 5]
```diff
@@ -114,6 +124,96 @@ boot_spd_initaep_get(spdid_t spdid)
...
+{
+	PRINTLOG(PRINT_ERROR, "in invstk overflow and underflow fault handler, fault happens in component:%u\n\n", (unsigned int)cos_inv_token());
+	fault_reg_print(cos_inv_token());
+	return;
+}
+
+void
+fault_comp_not_exist(unsigned long sp, unsigned long ip, unsigned long fault_addr, unsigned long fault_type)
+{
+	PRINTLOG(PRINT_ERROR, "in component does not exist fault handler, fault happens in component:%u\n\n", (unsigned int)cos_inv_token());
```
Comment (phanikishoreg) [inline src/components/implementation/no_interface/llbooter/boot_deps.h]: This applies to all PRINT statements, grammar doesn't seem right!
I'd rephrase it as: "\"Component Does Not Exist\" fault happened in component [%u]" or something like that.

File: src/components/interface/fault_handler/fault_handler.h [hunk 6]
```diff
@@ -0,0 +1,12 @@
+#ifndef FAULT_HANDLER_H
+#define FAULT_HANDLER_H
+
+void fault_div_by_zero(unsigned long sp, unsigned long ip, unsigned long fault_addr, unsigned long fault_type);
```
Comment (phanikishoreg) [inline src/components/interface/fault_handler/fault_handler.h]: Yeah, I'd use `word_t` where possible.

File: src/kernel/capinv.c [hunk 7]
```diff
@@ -1294,7 +1294,7 @@ static int __attribute__((noinline)) composite_syscall_slowpath(struct pt_regs *
 			vaddr_t    entry_addr    = __userregs_get3(regs);
 			invtoken_t token         = __userregs_get4(regs);
 
-			ret = sinv_activate(ct, cap, capin, dest_comp_cap, entry_addr, token);
+			ret = sinv_activate(ct, cap, capin, dest_comp_cap, CAP_SINV, entry_addr, token);
```
Comment (phanikishoreg) [inline src/kernel/capinv.c]: `sinv_activate` with a type? I think you should leave `sinv_activate` as is. In the FAULTACTIVATE, use `faultinv_activate` or something and call an internal function that would be `sinv_activate_type` within `inv.h`

File: src/kernel/include/inv.h [hunk 8]
```diff
@@ -286,7 +286,7 @@ arcv_introspect(struct cap_arcv *r, unsigned long op, unsigned long *retval)
  */
 
 static inline void
-sinv_call(struct thread *thd, struct cap_sinv *sinvc, struct pt_regs *regs, struct cos_cpu_local_info *cos_info)
+sinv_call(struct thread *thd, struct cap_sinv *sinvc, struct pt_regs *regs, struct cos_cpu_local_info *cos_info, unsigned long in_fault)
```
Comment (phanikishoreg) [inline src/kernel/include/inv.h]: Just curious why `unsigned long in_fault`? Wish we could use `bool` here. :-P

File: src/kernel/include/shared/cos_types.h [hunk 9]
```diff
@@ -130,6 +131,7 @@ typedef enum {
 typedef enum {
 	CAP_FREE = 0,
 	CAP_SINV,       /* synchronous communication -- invoke */
+	CAP_FLT,        /* communication of fault handler -- invoke*/
```
Comment (phanikishoreg) [inline src/kernel/include/shared/cos_types.h]: `communication of fault handler`? Not sure what that means.

File: src/kernel/include/shared/cos_types.h [hunk 10]
```diff
@@ -222,17 +225,21 @@ captbl_idsize(cap_t c)
...
- * 8-11 = our component,
- * 12-13 = vm pte for booter
- * 14-15 = untyped memory pgtbl root,
- * 16-17 = vm pte for physical memory,
- * 18-19 = km pte,
- * 20-21 = comp0 captbl,
- * 22-23 = comp0 pgtbl root,
- * 24-27 = comp0 component,
- * 28~(20+2*NCPU) = per core alpha thd
+ * 4-7 = memory access fault handler,
```
[the remark this answers is not in the document]
Comment (phanikishoreg) [reply src/kernel/include/shared/cos_types.h]: :-(  Perhaps a comment somewhere on what `memory access` fault handler is and what it generalizes would be useful. I'm still making my guesses here and that probably is not right!

File: src/kernel/include/shared/cos_types.h [hunk 11]
```diff
@@ -287,11 +311,27 @@ enum
...
 };
 
-enum
-{
+typedef enum {
 	/* thread id */
 	THD_GET_TID,
-};
+	/* get regs */
+	THD_GET_FAULT_REG0,
```
Comment (phanikishoreg) [inline src/kernel/include/shared/cos_types.h]: Are fault regs different from regular regs? Even if they represent a "fault" context and not regular context, I think the registers are still same. My point is, `FAULT_IP == THD.FAULT.IP` in some sense. Perhaps instead of having different set of enums here, have some kind of `fault_introspect` that passes an in_fault flag and fetches fault_regs?

Comment (phanikishoreg) [reply src/kernel/include/shared/cos_types.h]: Also I don't think we should have `REGX` notation for `IP` and `SP` at the very least?

File: src/platform/i386/exception.c [hunk 12]
```diff
@@ -13,19 +13,47 @@ print_regs_state(struct pt_regs *regs)
 {
 	PRINTK("registers:\n");
 	PRINTK("General registers-> EAX: %x, EBX: %x, ECX: %x, EDX: %x\n", regs->ax, regs->bx, regs->cx, regs->dx);
-	PRINTK("Segment registers-> CS: %x, DS: %x, ES: %x, FS: %x, GS: %x, SS: %x\n", regs->cs, regs->ds, regs->es,
-	       regs->fs, regs->gs, regs->ss);
-	PRINTK("Index registers-> ESI: %x, EDI: %x, EIP: %x, ESP: %x, EBP: %x\n", regs->si, regs->di, regs->ip,
-	       regs->sp, regs->bp);
+	PRINTK("Segment registers-> CS: %x, DS: %x, ES: %x, FS: %x, GS: %x, SS: %x, ESI: %x, EDI: %x \n", regs->cs, regs->ds, regs->es,
+	       regs->fs, regs->gs, regs->ss, regs->si, regs->di);
+	PRINTK("Index registers-> EIP: %x, ESP: %x, EBP: %x\n", regs->ip, regs->sp, regs->bp);
```
Comment (phanikishoreg) [inline src/platform/i386/exception.c]: Wait, aren't `SI` and `DI` `Index registers`? Why this change?

File: src/platform/i386/exception.c [hunk 13]
```diff
@@ -13,19 +13,47 @@ print_regs_state(struct pt_regs *regs)
...
+fault_handler_sinv(struct pt_regs *regs, capid_t cap)
+{
+	struct cos_cpu_local_info *ci       = cos_cpu_local_info();
+	struct thread *            curr_thd = thd_current(ci);
+	struct cap_header *        fh;
+	struct comp_info *         cos_info;
+	thdid_t                    thdid = curr_thd->tid;
+	unsigned long              ip, sp;
+	u32_t                      fault_addr = 0, errcode, eip;
+
```
Comment (phanikishoreg) [inline src/platform/i386/exception.c]: I think before you `SINV` into the fault-handler, we should first see if it is a kernel-level fault or user-level fault. I don't think you'd succeed, perhaps leads to double-fault if you tried anything intelligent if it is a kernel-level fault (obviously it depends on the type of fault.) but I think there is no way a user-level handler would handle kernel-level faults (bugs) obviously. Right?

Comment (phanikishoreg) [reply src/platform/i386/exception.c]: I might be wrong here about "obviously it depends on the type of fault". 
I need to read myself but will the user-level be able to resolve any kind of kernel-level fault, even just a page-fault? I doubt it! 
In my opinion, all kernel level faults are our kernel `BUGS` in someway or the other.
